gc/base/SparseAddressOrderedFixedSizeDataPool.cpp

```diff
@@ -0,0 +1,423 @@
+	J9ObjSparseDataTableEntry lookupEntry;
+
+	lookupEntry.dataPtr = dataPtr;
+	return (J9ObjSparseDataTableEntry *)hashTableFind(_objectToSparseDataTable, &lookupEntry);
+}
+
+uintptr_t
+MM_SparseAddressOrderedFixedSizeDataPool::getDataSizeForDataPtr(void *dataPtr)
+{
+	J9ObjSparseDataTableEntry *entry = getJ9ObjSparseDataTableEntry(dataPtr);
+
+	if ((NULL == entry) || (entry->dataPtr != dataPtr)) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_failure(dataPtr);
+		return 0;
+	}
+
+	Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_success(dataPtr);
+	return entry->size;
+}
+
+void *
+MM_SparseAddressOrderedFixedSizeDataPool::getProxyObjectPtrForDataPtr(void *dataPtr)
+{
+	J9ObjSparseDataTableEntry *entry = getJ9ObjSparseDataTableEntry(dataPtr);
+
+	if ((NULL == entry) || (entry->dataPtr != dataPtr)) {
+		Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_failure(dataPtr);
+		return NULL;
+	}
+
+	Trc_MM_SparseAddressOrderedFixedSizeDataPool_findEntry_success(dataPtr);
+	return entry->proxyObjPtr;
+}
+
+bool
+MM_SparseAddressOrderedFixedSizeDataPool::isValidDataPtr(void *dataPtr)
+{
+	J9ObjSparseDataTableEntry *entry = getJ9ObjSparseDataTableEntry(dataPtr);
+	return ((entry != NULL) && (entry->dataPtr == dataPtr));
+}
+
+MM_SparseHeapLinkedFreeHeader *
+MM_SparseAddressOrderedFixedSizeDataPool::createNewSparseHeapFreeListNode(void *dataAddr, uintptr_t size)
+{
+	MM_SparseHeapLinkedFreeHeader *newFreeListNode = (MM_SparseHeapLinkedFreeHeader *)pool_newElement(_freeListPool);
+	if (NULL != newFreeListNode) {
+		newFreeListNode->setAddress(dataAddr);
+		newFreeListNode->setSize(size);
+		newFreeListNode->_next = NULL;
+		_freeListPoolFreeNodesCount++;
+	}
+
+	return newFreeListNode;
+}
+
+void *
+MM_SparseAddressOrderedFixedSizeDataPool::findSparseHeapFreeListEntryForData(uintptr_t size)
+{
+	Assert_MM_true(_freeListPoolFreeNodesCount > 0);
+	MM_SparseHeapLinkedFreeHeader *previous = NULL;
+	MM_SparseHeapLinkedFreeHeader *current = _heapFreeList;
+	void *returnAddr = NULL;
+	uintptr_t currSize = 0;
+
+	/* Find big enough free space */
+	while(NULL != current) {
+		if (current->_size >= size) {
+			break;
+		}
+		previous = current;
+		current = current->_next;
+	}
+
+	/* Impossible for current to be NULL since sparse heap will always have available space no matter what */
+	Assert_MM_true(NULL != current);
+
+	currSize = current->_size;
+	returnAddr = current->_address;
+	if (currSize == size) {
+		/* Remove current node from FreeList since we're using all of it */
+		if (NULL == previous) {
+			_heapFreeList = current->_next;
+		} else {
+			previous->_next = current->_next;
+		}
+		pool_removeElement(_freeListPool, current);
+		_freeListPoolFreeNodesCount--;
+	} else {
+		/* Update current entry */
+		current->setAddress((void*)((uintptr_t)returnAddr + size));
+		current->contractSize(size);
+		/* Update largest free entry */
+		if (_largestFreeEntryAddr == returnAddr) {
+			_approxLargestFreeEntry -= size;
```

Comment:
hm, wondering how accurate this will be over long run, and why is this needed anyway!?
but let's leave it for now.